release/cibuild.py

```diff
@@ -10,163 +10,254 @@
+        raise BuildError("Unsupported platform: %s" % self.system)
+
+    @property
+    def release_dir(self):
+        return os.path.join(self.root_dir, "release")
+
+    @property
+    def build_dir(self):
+        return os.path.join(self.release_dir, "build")
+
+    @property
+    def dist_dir(self):
+        return os.path.join(self.release_dir, "dist")
+
+    @property
+    def docker_tag(self):
+        if self.branch == "master":
+            return "dev"
+        return self.version
+
+    def archive(self, path):
+        # ZipFile and tarfile have slightly different APIs. Fix that.
+        if self.system == "Windows":
+            a = zipfile.ZipFile(path, "w")
+            a.add = a.write
+            return a
+        else:
+            return tarfile.open(path, "w:gz")
+
+    def archive_name(self, bdist: str) -> str:
+        if self.system == "Windows":
+            ext = "zip"
+        else:
+            ext = "tar.gz"
+        return "{project}-{version}-{platform}.{ext}".format(
+            project=bdist,
+            version=self.version,
+            platform=self.platform_tag,
+            ext=ext
+        )
+
+    @property
+    def bdists(self):
+        ret = {
+            "mitmproxy": ["mitmproxy", "mitmdump", "mitmweb"],
+            "pathod": ["pathoc", "pathod"]
+        }
+        if self.system == "Windows":
+            ret["mitmproxy"].remove("mitmproxy")
+        return ret
+
+    @property
+    def should_upload_docker(self) -> bool:
+        return all([
+            (self.tag or self.branch == "master"),
+            self.should_build_docker,
+            self.has_docker_creds,
+        ])
 
-def build_wheel():
+    @property
+    def should_upload_pypi(self) -> bool:
+        if self.tag and self.should_build_wheel and self.has_twine_creds:
+            return True
+        return False
+
+    def dump_info(self, fp=sys.stdout):
+        print("BUILD PLATFORM_TAG=%s" % self.platform_tag, file=fp)
+        print("BUILD ROOT_DIR=%s" % self.root_dir, file=fp)
+        print("BUILD RELEASE_DIR=%s" % self.release_dir, file=fp)
+        print("BUILD BUILD_DIR=%s" % self.build_dir, file=fp)
+        print("BUILD DIST_DIR=%s" % self.dist_dir, file=fp)
+        print("BUILD BDISTS=%s" % self.bdists, file=fp)
+        print("BUILD TAG=%s" % self.tag, file=fp)
+        print("BUILD BRANCH=%s" % self.branch, file=fp)
+        print("BUILD VERSION=%s" % self.version, file=fp)
+        print("BUILD UPLOAD_DIR=%s" % self.upload_dir, file=fp)
+
+
+def build_wheel(be: BuildEnviron):  # pragma: no cover
     click.echo("Building wheel...")
     subprocess.check_call([
         "python",
         "setup.py",
         "-q",
         "bdist_wheel",
-        "--dist-dir", DIST_DIR,
+        "--dist-dir", be.dist_dir,
     ])
-
-    whl = glob.glob(join(DIST_DIR, 'mitmproxy-*-py3-none-any.whl'))[0]
+    whl = glob.glob(os.path.join(be.dist_dir, 'mitmproxy-*-py3-none-any.whl'))[0]
```

Comment:
Should we pullout the wheel filename into a property.

Reply:
I wouldn't think it's needed - it's a static glob pattern that never changes and is not computed from the environment.